aptos-move/block-executor/src/scheduler_v2.rs

```diff
@@ -4,21 +4,366 @@
+                if *stored_successful_abort_incarnation < invalidated_incarnation {
+                    // A previous invalidation targeted an older incarnation of `invalidated_txn_idx`
+                    // which was successfully aborted and recorded.
+                    // Now, a newer incarnation of the same `invalidated_txn_idx` is being targeted.
+                    // This is an error: `SchedulerV2::finish_execution` must consume the AbortManager
+                    // instance for the `stored_successful_abort_incarnation` before an attempt to
+                    // abort a higher incarnation of the same `invalidated_txn_idx` can be made.
+                    return Err(code_invariant_error(format!(
+                        "Lower incarnation {} than {} already invalidated by Abort Manager for txn version ({}, {})",
+                        *stored_successful_abort_incarnation, invalidated_incarnation,
+                        self.owner_txn_idx, self.owner_incarnation
+                    )));
+                }
+                // If *stored_incarnation >= invalidated_incarnation, it means either the same
+                // or a newer incarnation (compared to the current invalidation) has already been
+                // successfully aborted by this AbortManager instance. This can happen because
+                // the reads from outdated incarnations are not assumed to be (eagerly) cleared.
+                // In such cases, no new abort action is needed for this specific call.
+                false
+            },
+        };
+
+        if action_needed {
+            let abort_outcome: Option<Incarnation> =
+                self.start_abort(invalidated_txn_idx, invalidated_incarnation)?;
+
+            // Update self.invalidations with the outcome. This will either insert a new entry
+            // or update an existing one from None to Some(outcome).
+            self.invalidations
+                .insert(invalidated_txn_idx, abort_outcome);
+        }
+
+        Ok(())
+    }
+
+    /// Attempts to initiate an abort for the given transaction and incarnation via the scheduler.
+    ///
+    /// Returns:
+    /// - `Ok(Some(incarnation))` if `[SchedulerV2::start_abort]` was successful for the given `incarnation`.
+    /// - `Ok(None)` if `[SchedulerV2::start_abort]` returned false (e.g., already aborted).
+    /// - `Err(PanicError)` if `[SchedulerV2::start_abort]` itself returns an error.
+    fn start_abort(
+        &self,
+        txn_idx: TxnIndex,
+        incarnation: Incarnation,
+    ) -> Result<Option<Incarnation>, PanicError> {
+        fail_point!("abort-manager-start-abort-none", |_| Ok(None));
+        fail_point!("abort-manager-start-abort-some", |_| Ok(Some(incarnation)));
+        Ok(self
+            .scheduler
+            .start_abort(txn_idx, incarnation)?
+            .then_some(incarnation))
+    }
+
+    // Returns an iterator over invalidated transaction indices, as well as full versions,
```

Comment:
Do you intend to return an iterator, or this comment is old, as now a map is returned?

Reply:
Good catch: outdated comment. Updated to say if it contains incarnation, then finish_abort still needs to be performed.